### PR TITLE
Add Azure Monitoring Module to enable default alert criteria and notifications

### LINKED ISSUE #86: Azure monitoring module
### Description
As a developer, I can route metrics (a signal type) from an **Azure App Service Plan** (cobalt-azure-simple-sp) using Azure Monitoring to trigger alerts based on minimal criteria defined from an alert rule.

### Acceptance Criteria
Reference: [Done-Done Checklist] (https://github.com/Microsoft/code-with-engineering-playbook/blob/master/Engineering/BestPractices/DoneDone.md)
- [x] A module level readme that reveals minimal configuration attributes for the Azure monitoring module.
- [x] A main.tf file clearly dedicated to using **azurerm_monitor_metric_alert** as a resource.
- [x] An Azure Monitoring module that derives it's App Service Plan value from the Azure Simple Template.  An App Service Plan can have many App Services. An App Service can not have many App Service plans.
- [x] A screenshot of Azure Monitoring alert rules displayed from within the Azure portal. Configuration should reflect an alert rule that takes action based on a metric of concern emitted from **Azure App Service Plan** (cobalt-azure-simple-sp).

### Resources
Azure Monitor Overview - https://docs.microsoft.com/en-us/azure/azure-monitor/overview
Azure Monitor Alerts Overview - https://docs.microsoft.com/en-us/azure/azure-monitor/platform/alerts-overview
Azure Monitor Alerts ARM Template - https://docs.microsoft.com/en-us/azure/azure-monitor/platform/alerts-metric-create-templates
Azure Monitor Supported Metrics - https://docs.microsoft.com/en-us/azure/azure-monitor/platform/metrics-supported

### Tasks

1. Validate and discuss what required variables and resources are needed as configurable attributes for the azure monitoring metric resource. (Is the cobalt-azure-simple-sp enough? Should we setup rules for the other resources like the gateway, vnet or app insights? What about smart groups?)

2. Decide which variables obtain their values from outside the azure monitoring module. (What variables should the Monitoring Module expose to the azure simple template? (ex. resource group name, resource group location))

3. Validate and discuss which modules will consume the output from the app monitoring module.

4. Choose which default/optional variables would be nice-to-haves.

5. Implement and test deployment of the Azure Monitoring Module via the Azure Simple Template.

### PR DESCRIPTION
**Pull request by TechnicallyWilliams**
## All Submissions:
-------------------------------------
* [YES] Have you followed the guidelines in our Contributing [document](./CONTRIBUTING.md)?
* [YES] Have you added an explanation of what your changes do and why you'd like us to include them?
* [YES] I have updated the documentation accordingly.
* [NO] I have added tests to cover my changes.
* [NA] All new and existing tests passed.
* [YES] My code follows the code style of this project.
* [YES] I ran lint checks locally prior to submission.
* [YES] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

## What is the current behavior?
-------------------------------------
- Cobalt templates currently have the option to monitor metrics at the individual app service level by leveraging the **App-Insights** module. However, this new module introduces a starting point for monitoring metrics emitted by other Azure resources. (ex. App Service Plan, App Gateway, Traffic Manager, etc.)

Issue Number: https://github.com/microsoft/cobalt/issues/86
Fixes #86 

## What is the new behavior?
-------------------------------------
- The new module introduced in this PR deploys the **Azure Monitor** Service. It is a custom Terraform module called **App-Monitoring** that Cobalt Templates can use in order to assign configurable **_metric alert criteria_** to individual Azure Resources for monitoring. 

- In addition to the **_metric alert criteria_**, this module introduces configurable **_action groups_** that when paired with **_metric alert criteria_** can trigger event types like e-mail notifications.

- In conclusion, the **App-Monitoring** module is now being referenced by the **Azure Simple Template** which now adds the following default behavior to the template:
   - The deployment of the **Azure Monitor** Service depends on if a template provides an e-mail address to the *App-Monitoring** module. Otherwise, the **App-Monitoring** module deploys nothing.
  -   **Azure Simple Template** default is now set up to deploy CPU resource monitoring of the **App Service Plan** Service.

## Does this introduce a breaking change?
-------------------------------------
- [NO]

## Any relevant logs, error output, etc?
-------------------------------------
image

## Other information
-------------------------------------

Mapping metric alert criteria to a resource:
image

### CODE REVIEW

File: infra/modules/providers/azure/app-monitoring/README.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
+# Azure Monitor
+
+This custom Terraform based `app-monitoring` module deploys the "Azure Monitor" service in order to offer visibility into the behavior of your deployed Azure Resources. This module is recommended for monitoring Azure Resources deployed alongside App Services (ex. App Service Plan, App Gateway, VM, etc.). For direct monitoring of an App Service, the [app-insights](https://github.com/microsoft/cobalt/tree/master/infra/modules/providers/azure/app-insights) module is instead recommended. That module leverages an "Azure Monitor" service called [application insights](https://www.terraform.io/docs/providers/azurerm/r/application_insights.html).
```
Comment (nmiodice) [inline infra/modules/providers/azure/app-monitoring/README.md]: A few nits, but overall this is a fantastic intro to the module!
 - Since everything in the repo is custom, I think we can omit that word
 - You can use a relative link to the app-insights module like so: `[app-insights](../../path-to-module)` (the path is made up here). That way it will reflect any changes in your PR if you decide to change that module, and will hold up if the repo name ever changes or the module is moved to another repo

File: infra/modules/providers/azure/app-monitoring/README.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+In addition to the `metric alert criteria`, this module introduces a configurable `action group` that when paired with `metric alert criteria` can be configured to trigger events like e-mail notifications.
+
+#### More on Azure Monitoring
+
+"Azure Monitor maximizes the availability and performance of your applications by delivering a comprehensive solution for collecting, analyzing, and acting on telemetry from your cloud and on-premises environments. ..All data collected by Azure Monitor fits into one of two fundamental types, metrics and logs. Metrics are numerical values that describe some aspect of a system at a particular point in time. They are lightweight and capable of supporting near real-time scenarios. Logs contain different kinds of data organized into records with different sets of properties for each type." - Source: [Azure Monitor Overview](https://docs.microsoft.com/en-us/azure/azure-monitor/overview)
```
Comment (nmiodice) [inline infra/modules/providers/azure/app-monitoring/README.md]: `..All` >> `All`

File: infra/modules/providers/azure/app-monitoring/README.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+The list of `metrics` available for configuration depends entirely on the Azure Resource chosen to be monitored. Please visit [Supported Metrics](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/metrics-supported) for a complete list of metrics supported by the Azure Monitor service.
+
+#### More on Terraform Resources
```
Comment (nmiodice) [inline infra/modules/providers/azure/app-monitoring/README.md]: `Dependent Terraform Resources` may be more descriptive of the content below

File: infra/modules/providers/azure/app-monitoring/README.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+module "app_monitoring" {
+  source                            = "../../modules/providers/azure/app-monitoring"
+  resource_group_name               = "${module.service_plan.resource_group_name}"
+  resource_ids                      = ["${module.service_plan.app_service_plan_id}"]
+  action_group_name                 = "${var.action_group_name}"
+  action_group_email_receiver       = "${var.action_group_email_receiver}"
+  metric_alert_criteria_namespace   = "${var.metric_alert_criteria_namespace}"
+  metric_alert_name                 = "${var.metric_alert_name}"
```
Comment (nmiodice) [inline infra/modules/providers/azure/app-monitoring/README.md]: Can you provide examples of what these values (lines 73 -80) might be + a link to a document on Azure docs that describes their meaning?

File: infra/modules/providers/azure/app-monitoring/main.tf [hunk 5]
```diff
@@ -0,0 +1,36 @@
+resource "azurerm_monitor_action_group" "appmonitoring" {
+  count               = "${var.action_group_email_receiver == "" ? 0 : 1}"
```
Comment (nmiodice) [inline infra/modules/providers/azure/app-monitoring/main.tf]: nifty! That's a great way to do optional resources in TF. Never seen that before.

File: infra/modules/providers/azure/app-monitoring/main.tf [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    metric_namespace = "${var.metric_alert_criteria_namespace}"
+    metric_name      = "${var.metric_alert_criteria_name}"
+    aggregation      = "${var.metric_alert_criteria_aggregation}"
+    operator         = "${var.metric_alert_criteria_operator}"
+    threshold        = "${var.metric_alert_criteria_threshold}"
+
+    dimension {
+      name     = "Instance"
+      operator = "Include"
+      values   = ["*"]
```
Comment (nmiodice) [inline infra/modules/providers/azure/app-monitoring/main.tf]: [question for a curious reader] Not seeing much about what this means in the docs [here](https://www.terraform.io/docs/providers/azurerm/r/monitor_metric_alert.html). Do you have more context?

File: infra/templates/azure-simple/variables.tf [hunk 7]
```diff
@@ -49,6 +49,55 @@ variable "appgateway_http_listener_protocol" {
...
+variable "metric_alert_criteria_operator" {
+  description = "A logical operator used for building metric alert criteria."
+  type        = "string"
+  default     = "GreaterThan"
+}
+
+variable "metric_alert_criteria_threshold" {
+  description = "The criteria threshold value that activates the metric alert."
+  type        = "string"
+  default     = "10"
```
Comment (nmiodice) [inline infra/templates/azure-simple/variables.tf]: Is this a reasonable default? > 70% CPU seems alarming, but any lower is likely OK

Also what time period is this metric collected over?

File: infra/modules/providers/azure/app-monitoring/main.tf (hunk 5 above)
Comment (TechnicallyWilliams) [reply infra/modules/providers/azure/app-monitoring/main.tf]: Haha Erik introduced me.

File: infra/modules/providers/azure/app-monitoring/README.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+- [azurerm_monitor_metric_alert](https://www.terraform.io/docs/providers/azurerm/r/monitor_metric_alert.html)
+
+- [azurerm_monitor_action_group](https://www.terraform.io/docs/providers/azurerm/r/monitor_action_group.html)
+
+Cobalt modules used to deploy the `app-monitoring` module include the following:
+
+- [Service Plan Module](../service-plan)
+
+Variables used to define and deploy the `app-monitoring` module:
```
Comment (erikschlegel) [inline infra/modules/providers/azure/app-monitoring/README.md]: You can provide a link to the `variables.tf` file. This way we avoid reconciling changes across the readme and the variables definition.

File: infra/modules/providers/azure/app-monitoring/README.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+| `metric_alert_criteria_aggregation` | &check; | Template  | The calculation used for building metric alert criteria. |
+| `metric_alert_criteria_operator` | &check; | Template  | A logical operator used for building metric alert criteria. |
+| `metric_alert_criteria_threshold` | &check; | Template | The criteria threshold value that activates the metric alert. |
+| `scaling_name` | &check; | Module | Local var with a constant value of "Instance". |
+| `scaling_operator` | &check; | Module | Local var with a constant value of "Include". |
+| `scaling_values` | &check; | Template | App instance names made available from app service plan scaling options. |
+
+## Module Usage
+
+The `action group` variables are only configurable for sending e-mail notifications when `metric alert criteria` crosses a configured threshold value. The values used to define `metric alert criteria` are provided by a template. If a template does not provide an e-mail by satisfying the `action_group_email_receiver` variable, all other configuration values will be ignored and this module will not be present in the Terraform plan generated for deployment:
```
Comment (erikschlegel) [inline infra/modules/providers/azure/app-monitoring/README.md]: Providing an example to the `module "app_monitoring"` definition should be sufficient here.

File: infra/modules/providers/azure/app-monitoring/README.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+1. Choose an appropriate e-mail address to receive alert notifications.
+2. Choose an appropriate name for the 'action group' that will hold the e-mail address. (ex. "E-mail alert group")
+3. Apply the e-mail address and name using your chosen or newly created template.
+
+### Module Output
+
+Once the deployments are completed successfully, the output for the current module will be in the format mentioned below:
+
+```terraform
+Outputs:
```
Comment (erikschlegel) [inline infra/modules/providers/azure/app-monitoring/README.md]: This is how I documented the exported attributes for another module as a reference https://github.com/microsoft/cobalt/tree/erisch/templates/azure-simple/feature/azkeyvault/infra/modules/providers/azure/traffic-manager#attributes-reference

File: infra/templates/azure-simple/variables.tf (hunk 7 above)
Comment (erikschlegel) [reply infra/templates/azure-simple/variables.tf]: I feel defaulting our rules to a 10% CPU Average alert would result into a busy inbox for module users.

File: infra/modules/providers/azure/app-monitoring/README.md (hunk 1 above)
Comment (TechnicallyWilliams) [reply infra/modules/providers/azure/app-monitoring/README.md]: Relative path was a great suggestion. Line 14 is where the correction was made.

File: infra/modules/providers/azure/app-monitoring/README.md (hunk 2 above)
Comment (TechnicallyWilliams) [reply infra/modules/providers/azure/app-monitoring/README.md]: Latest removes extra periods.

File: infra/modules/providers/azure/app-monitoring/README.md (hunk 4 above)
Comment (TechnicallyWilliams) [reply infra/modules/providers/azure/app-monitoring/README.md]: A link to the Azure doc's is now provided in the sub-sections titled "Configuring **metric alert criteria**" and "Configuring **action group**"

File: infra/modules/providers/azure/app-monitoring/README.md (hunk 3 above)
Comment (TechnicallyWilliams) [reply infra/modules/providers/azure/app-monitoring/README.md]: Terraform Resources are now under a sub-section titled **Definition**.

File: infra/modules/providers/azure/app-monitoring/main.tf (hunk 6 above)
Comment (TechnicallyWilliams) [reply infra/modules/providers/azure/app-monitoring/main.tf]: These are options for scaling your alert metrics out to server instances available according to the chosen app-service plan subscription level. I prefixed the variables with "scaling" to provide more clarity.

Here's the link for more explanation.
https://docs.microsoft.com/en-us/azure/azure-monitor/platform/alerts-metric-overview#monitoring-at-scale-using-metric-alerts-in-azure-monitor

File: infra/templates/azure-simple/variables.tf (hunk 7 above)
Comment (TechnicallyWilliams) [reply infra/templates/azure-simple/variables.tf]: > I feel defaulting our rules to a 10% CPU Average alert would result into a busy inbox for module users.

Default is now at 50%. High enough not to 'flood' an inbox but low enough for a demonstration that the notifications are working.

File: infra/modules/providers/azure/app-monitoring/README.md (hunk 8 above)
Comment (TechnicallyWilliams) [reply infra/modules/providers/azure/app-monitoring/README.md]: Variables now live within an **Argument Reference** section.

File: infra/modules/providers/azure/app-monitoring/README.md (hunk 10 above)
Comment (TechnicallyWilliams) [reply infra/modules/providers/azure/app-monitoring/README.md]: I like your way of documenting it because it provides a short explanation of what the output actually is. I'll make the same change in this Readme.

File: infra/modules/providers/azure/app-monitoring/README.md (hunk 9 above)
Comment (TechnicallyWilliams) [reply infra/modules/providers/azure/app-monitoring/README.md]: This was removed in favor of the module "app_monitoring" definition in the form of a code block.

File: infra/templates/azure-simple/variables.tf (hunk 7 above)
Comment (TechnicallyWilliams) [reply infra/templates/azure-simple/variables.tf]: >

> Is this a reasonable default? > 70% CPU seems alarming, but any lower is likely OK
> 
> Also what time period is this metric collected over?

I added time frequency and time period variables with descriptions and default variables at the template level. ("metric_alert_frequency"; "metric_alert_period")
